### ISSUE TITLE
How to run migrations in parallel

### ISSUE THREAD

**Issue by amiralgazi** [lecaillon/Evolve#129]
Hi

Is there any way to run Evolve migrations in parallel for many databases?

I have a for loop on my databases list for which I run the migrations.
It works fine when I run it synchronously, but when I try to run the migrations in parallel I get the following exception:

Evolve.EvolveSqlException: Cannot release the application lock (Database Principal: 'public', Resource: 'Evolve') because it is not currently held. Sql query: sp_releaseapplock Cannot release the application lock (Database Principal: 'public', Resource: 'Evolve') because it is not currently held. ---> System.Data.SqlClient.SqlException: Cannot release the application lock (Database Principal: 'public', Resource: 'Evolve') because it is not currently held.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Evolve.Dialect.SQLServer.SQLServerDatabase.<>c.<ReleaseApplicationLock>b__11_1(IDbCommand cmd)
   at Evolve.WrappedConnectionEx.Execute[T](WrappedConnection wrappedConnection, String sql, Func`2 query, Action`1 setupDbCommand)
   --- End of inner exception stack trace ---
   at Evolve.WrappedConnectionEx.Execute[T](WrappedConnection wrappedConnection, String sql, Func`2 query, Action`1 setupDbCommand)
   at Evolve.Dialect.SQLServer.SQLServerDatabase.ReleaseApplicationLock()
   at Evolve.Evolve.InternalExecuteCommand(Action`1 commandAction)

**Comment by lecaillon** [comment]
By default Evolve will use a session level lock to coordinate the migrations on multiple nodes. This prevents two distinct Evolve executions from executing an Evolve command on the same database at the same time.

If your migrations are executed on a single node you can safely desable this behavior:
```c#
Evolve.EnableClusterMode = false;
```

**Comment by amiralgazi** [comment]
@lecaillon Thanks for your reply,

I have done what you had suggested, and now I get this exception:

Evolve.EvolveSqlException: There is already an object named 'changelog' in the database. Sql query: CREATE TABLE [dbo].[changelog] ( id INT NOT NULL IDENTITY(1,1) PRIMARY KEY, type TINYINT, version... There is already an object named 'changelog' in the database. ---> System.Data.SqlClient.SqlException: There is already an object named 'changelog' in the database.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToPipe, Int32 timeout, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Evolve.WrappedConnectionEx.Execute[T](WrappedConnection wrappedConnection, String sql, Func`2 query, Action`1 setupDbCommand)
   --- End of inner exception stack trace ---
   at Evolve.WrappedConnectionEx.Execute[T](WrappedConnection wrappedConnection, String sql, Func`2 query, Action`1 setupDbCommand)
   at Evolve.WrappedConnectionEx.ExecuteNonQuery(WrappedConnection wrappedConnection, String sql, Nullable`1 commandTimeout)
   at Evolve.Dialect.SQLServer.SQLServerMetadataTable.InternalCreate()
   at Evolve.Metadata.MetadataTable.InternalCreateIfNotExists()
   at Evolve.Metadata.MetadataTable.Execute[T](Func`1 func, Boolean createIfNotExists)
   at Evolve.Evolve.InternalMigrate(DatabaseHelper db)
   at Evolve.Evolve.InternalExecuteCommand(Action`1 commandAction)


I'll try to explain what I am trying to do.
Synchronously it works like this:

        public void RunMigrationOnAllDatabases()
        {
            IEnumerable<string> dbNames = FetchExistingDatabaseNames();
            foreach (string dbName in dbNames)
            {
                ProcessMigration(dbName);
            }
        }

        private void ProcessMigration(string dbName)
        {
            using (var connection = new SqlConnection(connectionString))
            {
                connection.Open();
                connection.ChangeDatabase(dbName);
                evolveHelper.Migrate(connection);
            }
        }


Now, instead of for loop within RunMigrationOnAllDatabases function, I would like to set this:
            Parallel.ForEach(dbNames, new ParallelOptions() { MaxDegreeOfParallelism = Environment.ProcessorCount } , ProcessMigration);

**Comment by lecaillon** [comment]
Evolve is not meant to run in parallel. It is not a use case it addresses normally. You can try of course but you will certainly face issues like it. All I can say is that you would have to create a new instance of Evolve for each database I guess. 

The error you talk about concerns the Evolve metadata table needed in each database. It seems, it tries to create it multiple times in the same database. 

So 2 advices:
- use a new instance of Evolve each time
- use a new dedicated **SqlConnection** in your "loop"

**Comment by amiralgazi** [comment]
I think I've already done these two advices

    public void RunMigrationOnAllDatabases()
    {
        IEnumerable<string> dbNames = FetchExistingDatabaseNames();
        foreach (string dbName in dbNames)
        {
            ProcessMigration(dbName);
        }
    }

    private void ProcessMigration(string dbName)
    {
        using (var connection = new SqlConnection(connectionString))
        {
            connection.Open();
            connection.ChangeDatabase(dbName);
            evolveHelper.Migrate(connection);
        }
    }

       public void Migrate(SqlConnection connection)
        {
                evolve = InitEvolve(connection);

                if (evolve != null)
                {
                    evolve.Migrate();
                }
        }

        private Evolve.Evolve InitEvolve(SqlConnection connection)
        {
            if (connection == null)
            {
                return null;
            }

            string dbName = connection.Database;
            IEnumerable<string> folderPaths = GetMigrationFolderNames(dbName);

            if (folderPaths.IsNullOrEmpty())
            {
                return null;
            }

            if (!ValidateScriptsFormat(folderPaths, dbName))
            {
                return null;
            }

            return new Evolve.Evolve(GlobalConstants.EVOLVE_SETTINGS_FILE_NAME, connection, m => logger.LogInformation(m))
            {
                Locations = folderPaths,
                IsEraseDisabled = true,
                OutOfOrder = true,
                EnableClusterMode = false
            };
        }


Do you have more advices?

**Comment by lecaillon** [comment]
So I guess it is not doable. Sorry. Again Evolve has not been developed with concurrency in mind. 

A another solution could be to use the [Evolve CLI](https://evolve-db.netlify.com/cli/) instead. No thread safety implied, because it will be a separate process. But do you really need that degree of parallelization ...? Only you can answer this. ;)

**Comment by amiralgazi** [comment]
I see. 

We have a production server with more than 400 databases who should be updated to the same schema in each release. Every time we run the Evolve module it takes at least two hours until the process is done, so we wanted to find a way to make it faster.
By the way, I've seen that it takes more than an hour to run the API's function `Migrate()` (for the whole process of 400 databases), even if there is no migration needed.

**Comment by sam123evers** [comment]
Hi

I was getting the same error as @amiralgazi, which was solved by adding

`EnableClusterMode = false,`

as @lecaillon suggested.  

Unfortunately, I have a new error which I cannot solve.

I believe my issue is that the naming convention/file lookup functionality is not working correctly.

I have two sample dbs set up for a test.  I modify one, do a schema compare, change the filename of the schema compare to the naming convention - V1_0_1__Add_Column.sql - and place the migration file in db/migrations at the root of the project.

The first time I run the app it works as expected:
-the migration runs and  a column is added to the correct table.  yay.

My issue is that when I run the app again I get this error:

Evolve.EvolveException: 'Error executing script: V1_0_1__Add_Column.sql after 11 ms. Column names in each table must be unique. Column name 'Email' in table 'dbo.Users' is specified more than once. Sql query: ALTER TABLE [dbo].[Users]     ADD [Email] VARCHAR (100) NULL; Column names in each table must be unique. Column name 'Email' in table 'dbo.Users' is specified more than once.'

This tells me that Evolve is trying to run the same script again.  As far I understand, the changelog is supposed to be referenced to see if a certain migration has already been run.  Am I understanding this correctly?

Please advise

**Comment by lecaillon** [comment]
Run Evolve in parallel is not supported. When you use `EnableClusterMode = false` Evolve does not use a database lock so you can run an another instance of Evolve at the same time... but at your own risk. 
Starting from now it is your responsability to ensure each instance of Evolve works on a different database with a different metadatable. Bottom line it is not recommanded.

For the error message you have, it seems the second time Evolve has not read the metadatable called  `changelog` by default. Check both, this table and the parameters you use to run Evolve (cf: `Evolve.MetadataTableSchema`, `Evolve.MetadataTableName`)

**Comment by sam123evers** [comment]
@lecaillon 

Thank you so much for your timely reply!  Unfortunately, after trying you suggestions I still get the same error message the second time I run the application. 

### Evolve instance configuration:

```cs
var evolve = new Evolve.Evolve(cnx, msg => _logger.LogInformation(msg))
 {
      EnableClusterMode = false,
      MetadataTableName = "changelog",
      MetadataTableSchema = "dbo",
      Locations = new[] { location },
      IsEraseDisabled = true,
      Placeholders = new Dictionary<string, string>
      {
             ["${table4}"] = "table4"
      }
 };
evolve.Migrate();
```

### SQL Object Explorer view(after migration script has been run):


![SQL_Object_Explorer](https://user-images.githubusercontent.com/25334488/70570494-9ee7f200-1b61-11ea-8666-67cac72a64c7.PNG)

### Migration Script ( V1_0_1__Add_Column.sql ):

```sql
GO
USE [Evolve_Before];

GO
PRINT N'Altering [dbo].[Users]...';

GO
ALTER TABLE [dbo].[Users]
    ADD [Email] VARCHAR (100) NULL;

GO
PRINT N'Update complete.';
GO
```
### changelog:

 
![changelog](https://user-images.githubusercontent.com/25334488/70571136-d7d49680-1b62-11ea-8367-6d872375302f.PNG)

We would really like this to work.  any suggestions are welcome. Thanks again for your time and assistance!

**Comment by lecaillon** [comment]
Very strange. Have you made the test from scratch the second time ? At least delete the line 2 and 3 of the changelog. 
And use a "standard" Evolve configuration if it suits you. Something like:
```
var evolve = new Evolve.Evolve(cnx, msg => _logger.LogInformation(msg))
 {
      Locations = new[] { location },
      IsEraseDisabled = true
 };
evolve.Migrate();
```
